JSF applist doesn't work with h2 anymore

From server log:
```
Caused by: org.h2.jdbc.JdbcSQLException: Order by expression "LOWER(RESOURCEEN0_.NAME)" must be in the result list in this case; SQL statement:
select distinct resourceen0_.id as id1_47_, resourceen0_.deletable as deletabl2_47_, resourceen0_.fcOwner as fcOwner3_47_, resourceen0_.localPortId as localPor4_47_, resourceen0_.name as name5_47_, resourceen0_.release_id as release_8_47_, resourceen0_.resourceGroup_id as resource9_47_, resourceen0_.resourceType_id as resourc10_47_, resourceen0_.softlinkId as softlink6_47_, resourceen0_.v as v7_47_ from TAMW_resource resourceen0_ left outer join TAMW_resourceType resourcety1_ on resourceen0_.resourceType_id=resourcety1_.id left outer join TAMW_consumedResRel consumedma2_ on resourceen0_.id=consumedma2_.masterResource_id left outer join TAMW_resource resourceen3_ on consumedma2_.slaveResource_id=resourceen3_.id where resourcety1_.name=? order by resourceen0_.deletable asc, lower(resourceen0_.name) asc limit ? [90068-173]
```

--- a/AMW_business/pom.xml
+++ b/AMW_business/pom.xml
@@ -191,7 +191,7 @@
 		<dependency>
 			<groupId>com.h2database</groupId>
 			<artifactId>h2</artifactId>
-			<version>1.3.168</version>
+			<version>1.4.197</version>
 			<scope>test</scope>
 		</dependency>
 		<!-- Needed for TestSetTest on Java 8 -->

--- a/AMW_business/src/main/java/ch/puzzle/itc/mobiliar/business/domain/applist/ApplistScreenDomainServiceQueries.java
+++ b/AMW_business/src/main/java/ch/puzzle/itc/mobiliar/business/domain/applist/ApplistScreenDomainServiceQueries.java
@@ -20,18 +20,13 @@
 
 package ch.puzzle.itc.mobiliar.business.domain.applist;
 
+import java.sql.SQLException;
 import java.util.List;
 
 import javax.inject.Inject;
 import javax.persistence.EntityManager;
 import javax.persistence.TypedQuery;
-import javax.persistence.criteria.CriteriaBuilder;
-import javax.persistence.criteria.CriteriaQuery;
-import javax.persistence.criteria.Join;
-import javax.persistence.criteria.JoinType;
-import javax.persistence.criteria.Predicate;
-import javax.persistence.criteria.Root;
-import javax.persistence.criteria.SetJoin;
+import javax.persistence.criteria.*;
 
 import ch.puzzle.itc.mobiliar.business.resourcegroup.entity.ResourceEntity;
 import ch.puzzle.itc.mobiliar.business.resourcegroup.entity.ResourceTypeEntity;
@@ -77,13 +72,31 @@ public class ApplistScreenDomainServiceQueries {
 
         q.distinct(true);
 
-        q.orderBy(cb.asc(appServer.get("deletable")), cb.asc(cb.lower(appServer.<String>get("name"))));
+        Expression<String> name = appServer.get("name");
+        // workaround for H2 incompatibility, see https://github.com/h2database/h2database/issues/408
+        if (!isH2()) {
+            name = cb.lower(name);
+        }
+
+        q.orderBy(cb.asc(appServer.get("deletable")), cb.asc(name));
         TypedQuery<ResourceEntity> query = entityManager.createQuery(q);
         if (maxResult != null) {
             query.setMaxResults(maxResult);
         }
 
         return query.getResultList();
+    }
+
+    private boolean isH2() {
+        org.hibernate.engine.spi.SessionImplementor sessionImp =
+                (org.hibernate.engine.spi.SessionImplementor) entityManager.getDelegate();
+        String databaseProductName = null;
+        try {
+            databaseProductName = sessionImp.connection().getMetaData().getDatabaseProductName();
+        } catch (SQLException e) {
+            e.printStackTrace();
+        }
+        return databaseProductName.equalsIgnoreCase("H2");
     }
 
 }

--- a/AMW_db_scripts/pom.xml
+++ b/AMW_db_scripts/pom.xml
@@ -22,7 +22,7 @@
 		<version.jboss.spec.javaee.6.0>1.0.7.Final</version.jboss.spec.javaee.6.0>
 
 		<!-- dependency versions -->
-		<version.h2>1.4.181</version.h2>
+		<version.h2>1.4.197</version.h2>
 		<version.pg>9.4-1200-jdbc41</version.pg>
 		<version.ora>11.2.0.3</version.ora>
 		

--- a/AMW_docker/configuration/wildfly_h2.cli
+++ b/AMW_docker/configuration/wildfly_h2.cli
@@ -1,6 +1,6 @@
 embed-server --std-out=echo
 
-#Oracle
+# H2
 data-source add --name=amwDS --driver-name="h2" --jndi-name=java:jboss/datasources/amwDS --user-name="${env.DB_USER:sa}" --use-java-context=true --connection-url="jdbc:h2:standalone/data/amw/amwFileDbIntegrationEmpty;DB_CLOSE_ON_EXIT=FALSE;DB_CLOSE_DELAY=-1;"
 data-source add --name=amwLiquibaseDS --driver-name="h2" --jndi-name=java:jboss/datasources/amwLiquibaseDS --user-name="${env.DB_USER:sa}" --use-java-context=true --connection-url="jdbc:h2:standalone/data/amw/amwFileDbIntegrationEmpty;DB_CLOSE_ON_EXIT=FALSE;DB_CLOSE_DELAY=-1;"
 

--- a/AMW_docker/configuration/wildfly_ora.cli
+++ b/AMW_docker/configuration/wildfly_ora.cli
@@ -1,8 +1,7 @@
 embed-server --std-out=echo
 
-#Oracle
+# Oracle
 data-source add --name=amwDS --driver-name="ojdbc6.jar" --driver-class="oracle.jdbc.OracleDriver" --jndi-name=java:jboss/datasources/amwDS --user-name="${env.DB_USER:MOBI_TRUNK}" --password="${env.DB_PASSWORD:MOBI_TRUNK_PW}" --use-java-context=true --connection-url="jdbc:oracle:thin:@//${env.DB_HOST:localhost}:${env.DB_PORT:1521}/xe"
-
 data-source add --name=amwLiquibaseDS --driver-name="ojdbc6.jar" --driver-class="oracle.jdbc.OracleDriver" --jndi-name=java:jboss/datasources/amwLiquibaseDS --user-name="${env.DB_USER:MOBI_TRUNK}" --password="${env.DB_PASSWORD:MOBI_TRUNK_PW}" --use-java-context=true --connection-url="jdbc:oracle:thin:@//${env.DB_HOST:localhost}:${env.DB_PORT:1521}/xe"
 
 # Mail

--- a/AMW_maia_federation/pom.xml
+++ b/AMW_maia_federation/pom.xml
@@ -220,7 +220,7 @@
         <dependency>
             <groupId>com.h2database</groupId>
             <artifactId>h2</artifactId>
-            <version>1.3.168</version>
+            <version>1.4.197</version>
             <scope>test</scope>
         </dependency>
 

